Add support for outputting Cobertura XML

rowillia opened the pull request:
[Cobertura](http://cobertura.github.io/cobertura/) is a popular code coverage tool who's output can be parsed by a large number of code coverage tools like Jenkins.

This is the same xml format that coveragepy uses so it's also got some precedence in the Python community.

gvanrossum commented on mypy/report.py:
```diff
@@ -326,6 +329,131 @@ def on_finish(self) -> None:
+        self.total_lines = 0
+        self.covered_lines = 0
+
+    def as_xml(self):
```
Alternatively make it a nested function inside add_packages()?

rowillia replied:
Can't make it nested as I need it in `on_finish`

gvanrossum commented:
- Lots of new code without coverage. Are there tests for the other reports?
- Maybe it's time to do a try: import... except... instead of all those nested imports?
- Could be worth it to update the lxml.etree stubs instead of adding type-ignores?

rowillia commented:
Thanks for the feedback @gvanrossum!

- None of the other reporters are tested AFAICT 😢 .  I Added in a set of unit tests and integration tests for this reporter.  The other reporters can probably piggy-back off of the new `outfile` field to write tests.
- @JukkaL What's your thoughts on better handling the `lxml` imports?  I remember in a previous conversation you had mentioned wanting to rip it out all together.
- Done - https://github.com/python/typeshed/pull/629

JukkaL commented:
My main concerns are not requiring `lxml` to do most things, and not using XSLT which I don't care about. Having report generators depend on `lxml` isn't terrible if they don't affect other parts of the codebase. A potential improvement would be to display something like `<whatever> not installed. Run 'python3 -m pip install <whatever>' to install it.` if we can't import `lxml`.

rowillia commented:
@JukkaL @gvanrossum Thanks for the feedback.  I refactored how `lxml` is handled with all the reporters to have a more helpful error message and updated typeshed as to not need the `type: ignore` comments.

gvanrossum commented:
Yay! Hope you don't mind I applied a few small fixups. I've merged.

Maybe a few Anys could be replaced with `etree._Element`? If you care can you submit a tiny follow-up PR?

gvanrossum commented:
Hm. Also, perhaps when the absence of lxml is detected it should not raise ImportError but just print the message and ignore the requested report?

dragongame commented:
Hello,
I would like to also use this very handy feature, however, team foundation server only accepts  [version coverage 04](https://github.com/cobertura/web/blob/master/htdocs/xml/coverage-04.dtd).
Are there any plans to update to the more current version instead of 03?

gvanrossum commented:
@dragongame I know relatively little about these things; maybe you can submit a PR to do what you want? (@rowillia Maybe you could review whatever @dragongame comes up with?)
